In [1]:
import sqlite3
import matplotlib.pyplot as plt
import pandas as pd
connection = sqlite3.connect('data_db.db')

# QUESTIONS:
- 1-How many customers are expected to churn and how many of them are not?
- 2-How many customers are having electricity, gas or both?
- 3-How is the contract type related churn?
- 4-Which channel sales has the most churn percentage
- 5-What is the average consumption for the last year?
- 6-what is the average forecasted consupmtion for the upcoming year?
- 7-what is the average price off peak , mid peak and peak for 2015?
- 8-what is the average forcasted price for the upcoming year?
- 9-what is the average net margin?
- 10-what is the correlation between the year and churn?
- 11-How is years of antiquity related to churn?

> We will answer these questions that might help us in visualzing and analyzing the data.



### 1-How many customers are expected to churn and how many of them are not?


In [2]:
sql="select count(id) from df where churn='0'"
cursor= connection.execute(sql)
cursor.fetchall()


[(13187,)]

In [3]:
sql="select count(id) from df where churn='1'"
cursor= connection.execute(sql)
cursor.fetchall()

[(1419,)]

### 2-How many customers are having electricity, gas or both?


In [4]:
sql="select count(id) from df where has_gas='t' and cons_12m = 0"
cursor= connection.execute(sql)
cursor.fetchall()

[(18,)]

In [5]:
sql="select count(id) from df where has_gas='f'"
cursor= connection.execute(sql)
cursor.fetchall()

[(11955,)]

In [6]:
sql="select count(id) from df where has_gas='t' and cons_12m > 0"
cursor= connection.execute(sql)
cursor.fetchall()

[(2633,)]

### 3-How is the contract type related churn?


In [7]:
sql="select count(id) from df where has_gas='t' and cons_12m > 0 and churn='1'"
cursor= connection.execute(sql)
cursor.fetchall()

[(216,)]

In [8]:
sql="select count(id) from df where has_gas='t' and cons_12m =0 and churn='1'"
cursor= connection.execute(sql)
cursor.fetchall()

[(1,)]

In [9]:
sql="select count(id) from df where has_gas='f' and cons_12m > 0 and churn='1'"
cursor= connection.execute(sql)
cursor.fetchall()

[(1193,)]

> Most customers who are expected to churn are consuming electricity only without gas.

### 4-Which channel sales has the most churned customers


In [10]:
sql="select channel_sales, count(churn) from df where churn='1' group by channel_sales"
cursor= connection.execute(sql)
cursor.fetchall()

[('MISSING', 283),
 ('ewpakwlliwisiwduibdlfmalxowmwpci', 75),
 ('foosdfpfkusacimwkcsosbicdxkicaua', 820),
 ('lmkebamcaaclubfxadlmueccxoimlema', 103),
 ('usilxuppasemubllopkaafesmlibmsdf', 138)]

### 5-What is the average electricity and gas consumption in the last 12 months?

In [11]:
sql="select avg(cons_12m), avg(cons_gas_12m) from df"
cursor= connection.execute(sql)
cursor.fetchall()

[(159220.2862522251, 28092.375325208817)]

### 6-what is the average forecasted consupmtion in the next 12 months?


In [12]:
sql="select avg(forecast_cons_12m) from df"
cursor= connection.execute(sql)
cursor.fetchall()

[(1868.614879501577,)]

### 7-what is the average price off peak , mid peak and peak for 2015?


In [13]:
sql="select avg(price_off_peak_var),avg(price_peak_var),avg(price_mid_peak_var) from price_data where price_date like '2015%' "
cursor= connection.execute(sql)
cursor.fetchall()

[(0.14102697259500283, 0.054630396898732635, 0.030496007458936835)]

### 8-what is the average forcasted price for the upcoming year?


In [14]:
sql="select avg(forecast_price_energy_off_peak),avg(forecast_price_energy_peak) from client_data"
cursor= connection.execute(sql)
cursor.fetchall()

[(0.1372832656853303, 0.05049076721895273)]

### 9-what is the average net margin for each year?


In [15]:
sql="select substr(date_activ,1,4), avg(net_margin),count(id) from df group by substr(date_activ,1,4)"
cursor= connection.execute(sql)
cursor.fetchall()

[('2003', 355.8551282051282, 39),
 ('2004', 351.9516788321169, 137),
 ('2005', 340.03582010582005, 189),
 ('2006', 363.45027777777784, 36),
 ('2007', 239.51526315789474, 95),
 ('2008', 232.9225, 124),
 ('2009', 133.75652044609677, 2690),
 ('2010', 173.9144472152951, 3609),
 ('2011', 184.18236040609128, 2758),
 ('2012', 218.2361285097194, 3704),
 ('2013', 219.95328431372522, 1224),
 ('2014', 307.89, 1)]

### 10-what is the correlation between the year and churn 


In [16]:
sql="select substr(date_activ,1,4),churn,count(id) from client_data where churn='1' group by substr(date_activ,1,4)"
cursor= connection.execute(sql)
cursor.fetchall()

[('2003', 1, 4),
 ('2004', 1, 12),
 ('2005', 1, 10),
 ('2006', 1, 2),
 ('2007', 1, 2),
 ('2008', 1, 6),
 ('2009', 1, 209),
 ('2010', 1, 259),
 ('2011', 1, 272),
 ('2012', 1, 456),
 ('2013', 1, 187)]

### 11-How is years of antiquity related to churn?

In [17]:
sql="select num_years_antig ,count(id) from df where churn='1' group by num_years_antig "
cursor= connection.execute(sql)
cursor.fetchall()

[(2, 3),
 (3, 329),
 (4, 472),
 (5, 200),
 (6, 339),
 (7, 42),
 (8, 5),
 (9, 1),
 (10, 4),
 (11, 12),
 (12, 11),
 (13, 1)]

# Conclusion:
- 1419 customers are expected to churn.
- Most of the customers are only having electricity contract type and a very small number of them are having gas contract type and so, most of the customers who are expected to churn are only consuming electricity.
- "foosdfpfkusacimwkcsosbicdxkicaua" channel sales have the greatest number of customers who are expected to churn.
- Average prices was calculated for 2015 only since we have the consumption data for each of the companies in it.
- Average net margin sarting from 2009 increased with a very huge rate.
- Contracts that got activated in 2012 have the highest expected churn number of customers.
- Almost 95% from customers who are expected to churn spent less than 6 years with bcg.
